lib/puppet/provider/user/useradd.rb

```diff
@@ -152,6 +152,46 @@ def groups=(value)
     set(:groups, value)
   end
 
+  def password=(value)
+    user = @resource[:name]
+
+    begin
+      # Puppet execute does not support strings as input, only files.
+      # The password is expected to be in an encrypted format given -e is specified:
+      tempfile = nil
+      tempfile = Tempfile.new("puppet_#{user}_pw", :encoding => Encoding::ASCII)
+      tempfile << "#{user}:#{value}\n"
+      tempfile.close()
+
+      # Options '-e' use encrypted password
+      # Must receive "user:enc_password" as input
+      # command, arguments = {:failonfail => true, :combine => true}
+      cmd = [command(:chpasswd), '-e']
+      execute_options = {
+        :failonfail => false,
+        :combine => true,
+        :stdinfile => tempfile.path,
+        :sensitive => has_sensitive_data?
+      }
+      output = execute(cmd, execute_options)
+
+      # chpasswd can return 1, even on success (at least on AIX 6.1); empty output
+      # indicates success
+      if output != ""
+        raise Puppet::ExecutionFailure, "chpasswd said #{output}"
+      end
+    rescue Puppet::ExecutionFailure  => detail
+      raise Puppet::Error, "Could not set password on #{@resource.class.name}[#{@resource.name}]: #{detail}", detail.backtrace
+    ensure
+      if tempfile
+        # Extra close will noop. This is in case the write to our tempfile
+        # fails.
+        tempfile.close()
+        tempfile.delete()
+      end
+    end
```

Comment:
The check for a nil `tempfile` can be removed if you structure the call so the Tempfile is created outside of the begin/rescue:

```ruby
tempfile = Tempfile.new(...)
begin
  ...
rescue
  tempfile.close
  tempfile.unlink
end